### PR TITLE
[Task-#3] Added controllers and Error handler

### PR DESCRIPTION
**Pull request by Arenooo**
Issue #3 Add controllers
---
Description
+ Added Base, Catalog, Basket and Order Controllers
+ Added Controller routing
+ Added ErrorHandler
+ Added Validators

### CODE REVIEW

File: WA.Pizza.API/Controllers/BasketController.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+using Microsoft.AspNetCore.Mvc;
+using WA.Pizza.Infrastructure.Data.Services;
+using WA.Pizza.Infrastructure.DTO.Basket;
+using WA.Pizza.Infrastructure.DTO.Catalog;
+
+namespace WA.Pizza.API.Controllers;
+
+public class BasketController : Controller
+{
+    private readonly BasketDataService _service;
```
Comment (dcwaas) [inline WA.Pizza.API/Controllers/BasketController.cs:11]: use more concrete naming, like _basketDataService

File: WA.Pizza.API/Controllers/BasketController.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    public BasketController(BasketDataService service)
+    {
+        _service = service;
+    }
+
+    [HttpGet("get/{id:int}")]
+    public async Task<BasketDTO> Get([Required] int id)
+    {
+        var item = await _service.Get(id);
+        return item;
```
Comment (dcwaas) [inline WA.Pizza.API/Controllers/BasketController.cs:22]: No need to declare item, return task directly

File: WA.Pizza.API/Controllers/BasketController.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    public async Task Delete([Required] int id)
+    {
+        await _service.RemoveItem(id);
+    }
+    
+    [HttpDelete("clear/{id:int}")]
+    public async Task Clear([Required] int id)
+    {
+        await _service.ClearBasket(id);
+    }
```
Comment (dcwaas) [inline WA.Pizza.API/Controllers/BasketController.cs:49]: get rid of async/await, throw task upper

File: WA.Pizza.API/Controllers/CatalogItemController.cs [hunk 4]
```diff
@@ -1,12 +1,51 @@
...
+using WA.Pizza.Infrastructure.Data.Services;
+using WA.Pizza.Infrastructure.DTO.Catalog;
 
-namespace WA.Pizza.API.Controllers
+namespace WA.Pizza.API.Controllers;
+
+public class CatalogItemController : ControllerBase
 {
-    public class CatalogItemController : Controller
+    private readonly CatalogDataService _service;
```
Comment (dcwaas) [inline WA.Pizza.API/Controllers/CatalogItemController.cs:11]: same issues like in basket controller
rename this controller to CatalogController

File: WA.Pizza.API/Controllers/OrderController.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+using WA.Pizza.Core.Models;
+using WA.Pizza.Infrastructure.Data.Services;
+using WA.Pizza.Infrastructure.DTO.Catalog;
+using WA.Pizza.Infrastructure.DTO.Order;
+
+namespace WA.Pizza.API.Controllers;
+
+public class OrderController : ControllerBase
+{
+    private readonly OrderDataService _service;
```
Comment (dcwaas) [inline WA.Pizza.API/Controllers/OrderController.cs:12]: same issues like in basket controller

File: WA.Pizza.Infrastructure/Validators/UpdateBasketItemRequestValidator.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+            .MinimumLength(1)
+            .WithMessage("Please enter a valid Name")
+            .MaximumLength(15)
+            .WithMessage("Your Name is too long");
+
+        RuleFor(i => i.Price)
+            .NotNull()
+            .WithMessage("Please enter a Price")
+            .GreaterThan(0)
+            .WithMessage("Please enter a valid Price");
```
Comment (dcwaas) [inline WA.Pizza.Infrastructure/Validators/UpdateBasketItemRequestValidator.cs:22]: Are you sure that Price and Name should be modified by adding items to basket?

File: WA.Pizza.Infrastructure/Validators/AddBasketItemRequestValidator.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+            .MinimumLength(1)
+            .WithMessage("Please enter a valid Name")
+            .MaximumLength(15)
+            .WithMessage("Your Name is too long");
+
+        RuleFor(i => i.Price)
+            .NotNull()
+            .WithMessage("Please enter a Price")
+            .GreaterThan(0)
+            .WithMessage("Please enter a valid Price");
```
Comment (dcwaas) [inline WA.Pizza.Infrastructure/Validators/AddBasketItemRequestValidator.cs:22]: Same question as for UpdateBasketItem

File: WA.Pizza.API/Extensions.cs [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+{
+    public static void ConfigureExceptionHandler(this IApplicationBuilder app)
+    {
+        app.UseExceptionHandler(appError =>
+        {
+            appError.Run(async context =>
+            {
+                context.Response.StatusCode = (int)HttpStatusCode.InternalServerError;
+                context.Response.ContentType = "application/json";
+                var contextFeature = context.Features.Get<IExceptionHandlerFeature>();
```
Comment (dcwaas) [inline WA.Pizza.API/Extensions.cs:19]: Add here switch case with your custom exception (set for it some message), and for others = get all inner exceptions messages to one ErrorMessage

File: WA.Pizza.API/Controllers/BasketController.cs (hunk 1 above)
Comment (Arenooo) [reply WA.Pizza.API/Controllers/BasketController.cs]: \+

File: WA.Pizza.API/Controllers/BasketController.cs (hunk 2 above)
Comment (Arenooo) [reply WA.Pizza.API/Controllers/BasketController.cs]: \+

File: WA.Pizza.API/Controllers/BasketController.cs (hunk 3 above)
Comment (Arenooo) [reply WA.Pizza.API/Controllers/BasketController.cs]: \+

File: WA.Pizza.API/Controllers/OrderController.cs (hunk 5 above)
Comment (Arenooo) [reply WA.Pizza.API/Controllers/OrderController.cs]: \+

File: WA.Pizza.Infrastructure/Validators/UpdateBasketItemRequestValidator.cs (hunk 6 above)
Comment (Arenooo) [reply WA.Pizza.Infrastructure/Validators/UpdateBasketItemRequestValidator.cs]: \+

File: WA.Pizza.Infrastructure/Validators/AddBasketItemRequestValidator.cs (hunk 7 above)
Comment (Arenooo) [reply WA.Pizza.Infrastructure/Validators/AddBasketItemRequestValidator.cs]: \+

File: WA.Pizza.API/Extensions.cs (hunk 8 above)
Comment (Arenooo) [reply WA.Pizza.API/Extensions.cs:19]: \+

**Review by dcwaas — APPROVED** [review]

File: WA.Pizza.API/Extensions.cs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+            });
+        });
+
+        Exception GetInnerException(Exception e)
+        {
+            while (e.InnerException != null)
+                e = e.InnerException;
+
+            return e;
+        }
```
Comment (var1ap-WA) [inline WA.Pizza.API/Extensions.cs:48]: do you have test on it?

Comment (Arenooo) [reply WA.Pizza.API/Extensions.cs:48]: it is

**Pull request merged**